clang/lib/Sema/BoundsAnalysis.cpp

```diff
@@ -136,128 +133,174 @@ bool BoundsAnalysis::AreDeclaredBoundsZero(const Expr *E, const Expr *V) {
   // 0.
   if (const auto *RBE = dyn_cast<RangeBoundsExpr>(E)) {
     if (const auto *BO = dyn_cast<BinaryOperator>(RBE->getUpperExpr())) {
-      auto *RHS = IgnoreCasts(BO->getRHS());
+      Expr *RHS = IgnoreCasts(BO->getRHS());
       if (const auto *Lit = dyn_cast<IntegerLiteral>(RHS)) {
-        auto *LHS = IgnoreCasts(BO->getLHS());
+        Expr *LHS = IgnoreCasts(BO->getLHS());
         return Lit->getValue().getLimitedValue() == 0 &&
-               Lexicographic(Ctx, nullptr).CompareExpr(LHS, V) ==
-               Lexicographic::Result::Equal;
+               Lex.CompareExpr(LHS, V) == Lexicographic::Result::Equal;
       }
     }
   }
   return false;
 }
 
-void BoundsAnalysis::FillGenSet(Expr *E,
-                                ElevatedCFGBlock *EB,
-                                ElevatedCFGBlock *SuccEB) {
+void BoundsAnalysis::HandlePointerDeref(UnaryOperator *UO,
+                                        ElevatedCFGBlock *EB,
+                                        ElevatedCFGBlock *SuccEB) {
 
-  // Handle if conditions of the form "if (*e1 && *e2)".
-  if (const auto *BO = dyn_cast<const BinaryOperator>(E)) {
-    if (BO->getOpcode() == BO_LAnd) {
-      FillGenSet(BO->getLHS(), EB, SuccEB);
-      FillGenSet(BO->getRHS(), EB, SuccEB);
-    }
-  }
+  const Expr *E = IgnoreCasts(UO->getSubExpr());
 
-  // Check if the edge condition contains a pointer deref.
-  if (!ContainsPointerDeref(E))
-    return;
+  // TODO: Handle accesses of the form:
+  // "if (*(p + i) && *(p + j) && *(p + k))"
 
-  E = IgnoreCasts(E);
+  // For conditions of the form "if (*p)".
+  if (const auto *D = dyn_cast<DeclRefExpr>(E)) {
```

Comment:
I think you need to check for the pattern:
```
   implicit lvalue-to-rvalue cast 
      declref
```
instead of relying on IgnoreCasts to strip out the lvalue-to-rvalue cast.